### ISSUE TITLE
NaN error while training Conformer-Transducer with automatic mixed precesion

### ISSUE THREAD

**Issue by qhoangdl** [NVIDIA/NeMo#3825]
Hi, thanks for the great toolkit!

I'm training the conformer-transducer-medium on librispeech-960hr from scratch, following the config in `examples/asr/conf/conformer/conformer_transducer_bpe.yaml`

I use a sentencepiece unigram tokenizer with a vocab size of 1024 generated using the script in NeMo. The batch size is set at 32 on each GPU for a global batch size of 256 (8 A100 GPUs are used). I use automatic mixed precision with amp_level O1. I tried a few different settings for the learning rate and weight decay to adjust for the batch size of 256, which is smaller than the default 2k batch size, and here are the results:

1. Learning rate = 2, weight decay = 1e-6. The training went well and got 2.8/7.6 WER on the dev-clean/dev-other sets at epoch 144. But suddenly got NaN error at epoch 145. I tried resuming the training from checkpoint but got NaN error again at the end of epoch 145.

2. Learning rate = 5, weight decay = 1e-3 (the default setting). The training ran for 25 epoch and got NaN error at epoch 26.

3. Learning rate = 2, weight decay = 1e-4. The training ran for 58 epochs and got NaN error at epoch 59.

Given that the NaN error occurred so late in the training, I doubt that the error was caused by some training example. So, any suggestions to find the root causes would be really appreciated!

Thank you!

**Comment by titu1994** [comment]
We are currently triaging AMP with Conformers, the attention layer in them seems to sometimes cause nans when computing the attention map. We would recommend training with FP 32 until we can verify and fix AMP. It may sometimes work by lowering LR and with some combinations of weight decay, but its taking a chance. 
@VahidooX  to add other relevant information.

**Comment by VahidooX** [comment]
Unfortunately the conformer is not fully stable with fp16 currently. I did not get the time to investigate and fix it yet. You may try gradient clipping of 1.0 or even lower maybe 0.3. But there is no guarantee to work. If you have A100 GPUs, you're in luck as you can use bf16 instead by setting precision=bf16. It works similar to fp16 while it is stable.

**Comment by qhoangdl** [comment]
thanks @titu1994  and @VahidooX  very much for your feedback! I tried resuming learning with precision = 32. I manage to keep batch size of 32 per gpu by reducing fused_batch_size to 8. It seems to avoid the NaN error. I'll try bf16 later to see if it works.

Btw, regarding to the WER, does 2.8/7.6 WER on dev-clean/dev-other at epoch 144 (about 144k update steps) look okay? Would it eventually get to 6.3% on test-other like you mentioned?

**Comment by VahidooX** [comment]
That is not bad, but I could get better results at that epoch with some specific configs. I used WPE for those initial experiments with no fast emit. Are you using fast emit? What are the lr and weight decay?

**Comment by qhoangdl** [comment]
hi @VahidooX , I thought that sentencepiece unigram is better than WPE. Should I change to WPE?

I'm not using fast emit. I'm setting lr = 2 and weight decay = 1e-6 for the batch size of 256. What config would you suggest? Should I use the default setting lr = 5 and weight decay = 1e-3?

Also, if I use bf16, should I use native or apex amp backend?

Thank you!

**Comment by VahidooX** [comment]
Mostly WPE and SPE unigram tokenizers should give similar results, sometimes results are better or worse in the range of 0.1 to 0.2% at most. You may also disable the use_start_end_token by setting it to False as it is not needed. For this model and dataset, you may keep WPE tokenizer if you want to continue the training but generally we recommend to use SPE unigram as it is more flexible.

I have not tried batch size of 256 but it should be similar. I suggest to try lr=5 and weight decay=1e-3. For bf16, you may use the native backend.
